api/src/metrics.rs

```diff
@@ -0,0 +1,53 @@
+// Copyright (c) The Diem Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use diem_metrics::{register_histogram_vec, HistogramVec};
+
+use once_cell::sync::Lazy;
+use warp::log::{custom, Info, Log};
+
+static HISTOGRAM: Lazy<HistogramVec> = Lazy::new(|| {
+    register_histogram_vec!(
+        "diem_api_requests",
+        "API requests latency grouped by method, operation_id and status",
+        &["method", "operation_id", "status"]
+    )
+    .unwrap()
+});
+
+static RESPONSE_STATUS: Lazy<HistogramVec> = Lazy::new(|| {
+    register_histogram_vec!(
+        "diem_api_response_status",
+        "API requests latency grouped by status code only",
+        &["status"]
+    )
+    .unwrap()
+});
+
+// Record metrics by method, operation_id and status.
+// The operation_id is the id for the request handler.
+// Should use same `operationId` defined in `openapi.yaml` whenever possible.
+pub fn metrics(operation_id: &'static str) -> Log<impl Fn(Info) + Copy> {
+    let func = move |info: Info| {
+        HISTOGRAM
+            .with_label_values(&[
+                info.method().to_string().as_str(),
+                operation_id,
+                info.status().as_u16().to_string().as_str(),
+            ])
+            .observe(info.elapsed().as_secs_f64());
+    };
+    custom(func)
+}
+
+// Record metrics by response status.
+// This is for understanding the overview of responses in case server
+// is overloaded by unknown reason.
+pub fn status_metrics() -> Log<impl Fn(Info) + Copy> {
+    let func = move |info: Info| {
+        RESPONSE_STATUS
+            .with_label_values(&[info.status().as_u16().to_string().as_str()])
+            .observe(info.elapsed().as_secs_f64());
+    };
+    custom(func)
+}
```

Comment:
I'm not sure you need this because I think in prometheus you can aggregate over just one field of the other histogram.  Or does this include errors that aren't covered by methods? (in which you could just put `unknown` in place if possible).

Maybe ask the question what's the insight we'll gain from recording the latency separate from per method.

Reply:
the first one `metrics` is used for each route, when the route and url path matched, a request is processed by the handler, the metrics is record.
But there are other http requests like invalid url path, e.g. `/abcd/not_found`, it will not hit any defined routes, thus we need this request per response status, and you can see I append this metrics to the end of `index` route which aggregates all routes defined.

The purpose of this is what I explained in the comment:
```
// This is for understanding the overview of responses in case server
// is overloaded by unknown reason.
```
It makes sure we don't have blind spots.